app/Http/Controllers/Auth/LoginController.php

```diff
@@ -36,161 +28,78 @@ class LoginController extends BaseController
+    protected function verifyLogin(Request $request)
     {
-        $allowSuspended = true;
-        $allowInactive = true;
-        $token = VatsimSSO::requestToken(route('auth-vatsim-sso'), $allowSuspended, $allowInactive);
-        if ($token) {
-            $key = $token->token->oauth_token;
-            $secret = $token->token->oauth_token_secret;
-            Session::put('credentials.vatsim-sso', compact('key', 'secret'));
-
-            return redirect()->to(VatsimSSO::sendToVatsim());
+        try {
+            $accessToken = $this->provider->getAccessToken('authorization_code', [
+                'code' => $request->input('code'),
+            ]);
+        } catch (IdentityProviderException $e) {
+            return redirect()->route('dashboard')->withError('Something went wrong, please try again.');
         }
-        // Check if there was a CURL error code
-        if (VATSIMSSO::error()['code']) {
-            Log::error('VATSIMSSO was unable to reach CERT. Code:'.VATSIMSSO::error()['code'].' Message:'.VATSIMSSO::error()['message']);
 
-            return redirect()->back()->withErrors(['connection' => "We were unable to contact VATSIM's certification service. Please try again later. If this persists, please contact Web Services."]);
+        $resourceOwner = json_decode(json_encode($this->provider->getResourceOwner($accessToken)->toArray()));
+
+        if (
+            ! $resourceOwner->data ||
+            ! $resourceOwner->data->cid ||
+            ! $resourceOwner->data->personal ||
+            ! $resourceOwner->data->personal->name_first ||
+            ! $resourceOwner->data->personal->name_last ||
+            ! $resourceOwner->data->personal->email ||
+            ! $resourceOwner->data ||
+            ! $resourceOwner->data->vatsim ||
+            ! $resourceOwner->data->oauth->token_valid === 'true'
+        ) {
+            return redirect()->route('dashboard')->withError('You cannot use our services unless you provide the relevant permissions upon login. Please try again.');
         }
-        throw new \Exception('SSO failed: '.VatsimSSO::error()['message']);
-    }
-
-    protected function setVatsimAuth($userId)
-    {
-        Auth::guard('vatsim-sso')->loginUsingId($userId);
-    }
 
-    protected function attemptSecondaryAuth()
-    {
-        $member = Auth::guard('vatsim-sso')->user();
-
-        if ($member->hasPassword()) {
-            return redirect()->route('auth-secondary');
-        }
+        $account = $this->completeLogin($resourceOwner, $accessToken);
 
-        $intended = Session::pull('url.intended', route('site.home'));
+        Auth::guard('vatsim-sso')->loginUsingId($account->id);
 
-        Auth::login(Auth::guard('vatsim-sso')->user(), true);
-
-        return redirect($intended);
+        return SecondaryLoginController::attemptSecondaryAuth($account);
     }
 
-    public function loginSecondary(Request $request)
+    protected function completeLogin($resourceOwner, $token)
```

Comment:
Can we be any more specific about the types of these two parameters?

Reply:
We can type hint object, can't be much more specific than that without some additional complication

d590ec1b